RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstElectronDeadECAL.cc

```diff
@@ -0,0 +1,106 @@
+/** \class TauDiscriminationAgainstElectronDeadECALBase
+ *
+ * Base class for producing PFTau and PATTau discriminators
+ *
+ * Flag tau candidates reconstructed near dead ECAL channels,
+ * in order to reduce e -> tau fakes not rejected by anti-e MVA discriminator
+ *
+ * The motivation for this flag is this presentation:
+ *   https://indico.cern.ch/getFile.py/access?contribId=0&resId=0&materialId=slides&confId=177223
+ *
+ * \authors Lauri Andreas Wendland,
+ *          Christian Veelken
+ *
+ *
+ *
+ */
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "RecoTauTag/RecoTau/interface/TauDiscriminationProducerBase.h"
+#include "RecoTauTag/RecoTau/interface/AntiElectronDeadECAL.h"
+
+// helper function retrieve the correct tau type name
+template <class TauType>
+std::string getTauTypeNameString(bool capitalise) {
+  // this generic one shoudl never be called.
+  throw cms::Exception("TauDiscriminationAgainstElectronDeadECALBase")
+      << "Unsupported TauType used. You must use either reco::PFTau or pat::Tau.";
+}
+// template specialiazation to get the correct (PF/PAT)Tau type names
+template <>
+std::string getTauTypeNameString<reco::PFTau>(bool capitalise) {
+  return capitalise ? "PFRecoTau" : "pfRecoTau";
+}
+template <>
+std::string getTauTypeNameString<pat::Tau>(bool capitalise) {
+  return capitalise ? "PATTau" : "patTau";
+}
+
+template <class TauType, class TauDiscriminator>
+class TauDiscriminationAgainstElectronDeadECALBase : public TauDiscriminationProducerBase<TauType, TauDiscriminator> {
```

Comment:
Why ```*DeadECALBase``` is it supposed to be used as a base class (it wasn't in the earlier version of this PR)?
Please drop -> ```DeadECAL```.

Reply:
In the earlier version it inherited from specialized versions (for `reco::PFTau` or `pat::Tau`) of template class `TauDiscriminationProducerBase<TauType, TauDiscriminator>` for a generic tau discriminant. Now the template is for `TauDiscriminationAgainstElectronDeadECAL` for any TauType, so one more sensible renaming (if mandatory) will be removal of `Base`.

Reply:
>  so one more sensible renaming (if mandatory) will be removal of `Base`.

that's what I meant, actually, but ended up typing in a bit confusing way. 
(we'll have ```TauDiscriminationAgainstElectronDeadECAL```)